In [3]:
from transformers import AutoModelForCausalLM, AutoTokenizer

### **1. GPT-2**

In [ ]:
model_name = "gpt2"
model = AutoModelForCausalLM.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name)

In [ ]:
input_text = "iran in year 1403"
input_ids = tokenizer.encode(input_text, return_tensors="pt")

output = model.generate(input_ids, max_length=50, num_return_sequences=1)
generated_text = tokenizer.decode(output[0], skip_special_tokens=True)

print("\n\ngenerated_text : ", generated_text)

In [12]:
print(model)

GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(50257, 768)
    (wpe): Embedding(1024, 768)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPT2Block(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2Attention(
          (c_attn): Conv1D(nf=2304, nx=768)
          (c_proj): Conv1D(nf=768, nx=768)
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D(nf=3072, nx=768)
          (c_proj): Conv1D(nf=768, nx=3072)
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=768, out_features=50257, bias=False)
)


#### طبق خروجی بالا gpt2 ویژگی های

vocabulary-size = 50257

GPT2Block ساختار مدل ⟸ 12 بلاک

میدانیم که خروجی مدل های علی در نهایت یک بردار به اندازه تعداد واژگان است. علتش هم این بود که خروجی مدل های علی یک توزیع احتمال روی واژگان دیکشنری است. برای اینکه خروجی به این شکل باشد، اخر شبکه مدل علی همیشه یک لایه خطی میگذارند که خروجی را به برداری با اندازه تعداد واژگان تبدیل کند. این لایه ای که راجبش صحبت میکنیم، لایه زیر است (اخرین لایه در بالا) :    

🟥 **(lm_head): Linear(in_features=768, out_features=50257, bias=False)**





### **2. GPT-Neo**

In [ ]:
from transformers import AutoModelForCausalLM, AutoTokenizer

model_name = "EleutherAI/gpt-neo-125M"  # نسخه سبک‌ تر
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(
    model_name,
    device_map="auto",          # خودش روی GPU/CPU پخش می‌کنه
    torch_dtype="auto",         # حافظه کمتر
)

In [ ]:
input_text = "a cute cat"
input_ids = tokenizer.encode(input_text, return_tensors="pt")

output = model.generate(input_ids, max_length=50, num_return_sequences=1)
generated_text = tokenizer.decode(output[0], skip_special_tokens=True)

print("\n\ngenerated_text : ", generated_text)

In [5]:
print(model)

GPTNeoForCausalLM(
  (transformer): GPTNeoModel(
    (wte): Embedding(50257, 768)
    (wpe): Embedding(2048, 768)
    (drop): Dropout(p=0.0, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPTNeoBlock(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPTNeoAttention(
          (attention): GPTNeoSelfAttention(
            (attn_dropout): Dropout(p=0.0, inplace=False)
            (resid_dropout): Dropout(p=0.0, inplace=False)
            (k_proj): Linear(in_features=768, out_features=768, bias=False)
            (v_proj): Linear(in_features=768, out_features=768, bias=False)
            (q_proj): Linear(in_features=768, out_features=768, bias=False)
            (out_proj): Linear(in_features=768, out_features=768, bias=True)
          )
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPTNeoMLP(
          (c_fc): Linear(in_features=768, out_features=3072, bias=True)
          (c_proj): Linear(in_fe

### **3. XLNet**

استفاده میکند Permutation Language Modeling این مدل از روش

In [ ]:
model_name = "xlnet-base-cased"
model = AutoModelForCausalLM.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name)

In [ ]:
input_text = "technology in iran"
input_ids = tokenizer.encode(input_text, return_tensors="pt")

output = model.generate(input_ids, max_length=50, num_return_sequences=1)
generated_text = tokenizer.decode(output[0], skip_special_tokens=True)

print("\n\ngenerated_text : ", generated_text)

In [30]:
print(model)

XLNetLMHeadModel(
  (transformer): XLNetModel(
    (word_embedding): Embedding(32000, 768)
    (layer): ModuleList(
      (0-11): 12 x XLNetLayer(
        (rel_attn): XLNetRelativeAttention(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (ff): XLNetFeedForward(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (layer_1): Linear(in_features=768, out_features=3072, bias=True)
          (layer_2): Linear(in_features=3072, out_features=768, bias=True)
          (dropout): Dropout(p=0.1, inplace=False)
          (activation_function): GELUActivation()
        )
        (dropout): Dropout(p=0.1, inplace=False)
      )
    )
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (lm_loss): Linear(in_features=768, out_features=32000, bias=True)
)


#### **معماری مدل**

است XLNetLayer کل معماری این مدل 12 لایه

هر لایه این مدل یک مکانیزم توجه + شبکه پیش خور است

XLNetLayer : RelativeAttention + FeedForward

vocab_size : 32000

طول بردار خروجی این مدل : 32000   

آخرین لایه خطی این مدل که بردار های خروجی تولید میکند :    

🟧 **(lm_loss): Linear(in_features=768, out_features=32000, bias=True)**


### **4. CTRL**

In [ ]:
from transformers import AutoModelForCausalLM, AutoTokenizer

model_name = "Salesforce/ctrl"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name,force_download=True)

In [ ]:
input_text = "a little story"
input_ids = tokenizer.encode(input_text, return_tensors="pt")

output = model.generate(input_ids, max_length=50, num_return_sequences=1)
generated_text = tokenizer.decode(output[0], skip_special_tokens=True)

print("\n\ngenerated_text : ", generated_text)

In [ ]:
print(model)